File: android/app/src/main/java/com/taskflow/app/util/DateUtils.kt
package com.taskflow.app.util

import java.util.Calendar
import java.util.concurrent.TimeUnit

object DateUtils {

    fun startOfDay(timestamp: Long): Long {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = timestamp
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        return calendar.timeInMillis
    }

    fun endOfDay(timestamp: Long): Long {
        return startOfDay(timestamp) + TimeUnit.DAYS.toMillis(1) - 1
    }

    fun isToday(timestamp: Long): Boolean {
        return startOfDay(timestamp) == startOfDay(System.currentTimeMillis())
    }

    fun isTomorrow(timestamp: Long): Boolean {
        val tomorrow = startOfDay(System.currentTimeMillis()) + TimeUnit.DAYS.toMillis(1)
        return startOfDay(timestamp) == tomorrow
    }

    fun isThisWeek(timestamp: Long): Boolean {
        val now = Calendar.getInstance()
        val dayOfWeek = now.get(Calendar.DAY_OF_WEEK)
        val diffToMonday = if (dayOfWeek == Calendar.SUNDAY) 6 else dayOfWeek - 2
        val weekStart = startOfDay(now.timeInMillis) - diffToMonday * TimeUnit.DAYS.toMillis(1)
        val weekEnd = weekStart + 7 * TimeUnit.DAYS.toMillis(1)
        return timestamp in weekStart until weekEnd
    }

    fun formatDate(timestamp: Long?): String {
        if (timestamp == null) return ""
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = timestamp
        val today = startOfDay(System.currentTimeMillis())
        val target = startOfDay(timestamp)
        val diff = (target - today) / TimeUnit.DAYS.toMillis(1)

        return when {
            diff == 0L -> "今天"
            diff == 1L -> "明天"
            diff == -1L -> "昨天"
            diff in 2..6 -> {
                val weekdays = arrayOf("周日", "周一", "周二", "周三", "周四", "周五", "周六")
                weekdays[calendar.get(Calendar.DAY_OF_WEEK) - 1]
            }
            else -> "${calendar.get(Calendar.MONTH) + 1}月${calendar.get(Calendar.DAY_OF_MONTH)}日"
        }
    }

    fun formatRelativeTime(timestamp: Long): String {
        val diff = System.currentTimeMillis() - timestamp
        val minutes = TimeUnit.MILLISECONDS.toMinutes(diff)
        val hours = TimeUnit.MILLISECONDS.toHours(diff)
        val days = TimeUnit.MILLISECONDS.toDays(diff)

        return when {
            minutes < 1 -> "刚刚"
            minutes < 60 -> "${minutes} 分钟前"
            hours < 24 -> "${hours} 小时前"
            days < 7 -> "${days} 天前"
            else -> formatDate(timestamp)
        }
    }

    fun formatChatTime(timestamp: Long): String {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = timestamp
        val now = Calendar.getInstance()
        val sameDay = isToday(timestamp)
        return if (sameDay) {
            String.format("%02d:%02d", calendar.get(Calendar.HOUR_OF_DAY), calendar.get(Calendar.MINUTE))
        } else {
            "${calendar.get(Calendar.MONTH) + 1}/${calendar.get(Calendar.DAY_OF_MONTH)}"
        }
    }

    fun groupTasksByDate(tasks: List<com.taskflow.app.domain.model.Task>): List<Pair<String, List<com.taskflow.app.domain.model.Task>>> {
        val today = mutableListOf<com.taskflow.app.domain.model.Task>()
        val tomorrow = mutableListOf<com.taskflow.app.domain.model.Task>()
        val thisWeek = mutableListOf<com.taskflow.app.domain.model.Task>()
        val later = mutableListOf<com.taskflow.app.domain.model.Task>()

        for (task in tasks) {
            val dueDate = task.dueDate
            if (dueDate == null) {
                later.add(task)
                continue
            }
            when {
                isToday(dueDate) -> today.add(task)
                isTomorrow(dueDate) -> tomorrow.add(task)
                isThisWeek(dueDate) -> thisWeek.add(task)
                else -> later.add(task)
            }
        }

        val result = mutableListOf<Pair<String, List<com.taskflow.app.domain.model.Task>>>()
        if (today.isNotEmpty()) result.add("今天" to today.sortedBy { it.order })
        if (tomorrow.isNotEmpty()) result.add("明天" to tomorrow.sortedBy { it.order })
        if (thisWeek.isNotEmpty()) result.add("本周" to thisWeek.sortedBy { it.dueDate ?: Long.MAX_VALUE })
        if (later.isNotEmpty()) result.add("更晚" to later.sortedBy { it.dueDate ?: Long.MAX_VALUE })
        return result
    }
}

File: src/utils/dateUtils.ts
import { Task, FilterType, TaskGroup } from '@/types/task';

const dayMs = 24 * 60 * 60 * 1000;

export function startOfDay(ts: number): number {
  const d = new Date(ts);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

export function isSameDay(a: number, b: number): boolean {
  return startOfDay(a) === startOfDay(b);
}

export function isToday(ts: number): boolean {
  return isSameDay(ts, Date.now());
}

export function isTomorrow(ts: number): boolean {
  return isSameDay(ts, Date.now() + dayMs);
}

export function isThisWeek(ts: number): boolean {
  const now = new Date();
  const day = now.getDay();
  const diff = day === 0 ? 6 : day - 1;
  const monday = startOfDay(now.getTime() - diff * dayMs);
  const nextMonday = monday + 7 * dayMs;
  return ts >= monday && ts < nextMonday;
}

export function formatDate(ts?: number): string {
  if (!ts) return '';
  const d = new Date(ts);
  const now = new Date();
  const today = startOfDay(now.getTime());
  const target = startOfDay(ts);
  const diff = Math.round((target - today) / dayMs);

  if (diff === 0) return '今天';
  if (diff === 1) return '明天';
  if (diff === -1) return '昨天';
  if (diff > 1 && diff < 7) {
    const weekdays = ['周日', '周一', '周二', '周三', '周四', '周五', '周六'];
    return weekdays[d.getDay()];
  }
  return `${d.getMonth() + 1}月${d.getDate()}日`;
}

export function formatRelativeTime(ts: number): string {
  const diff = Date.now() - ts;
  const minutes = Math.floor(diff / 60000);
  const hours = Math.floor(diff / 3600000);
  const days = Math.floor(diff / dayMs);

  if (minutes < 1) return '刚刚';
  if (minutes < 60) return `${minutes} 分钟前`;
  if (hours < 24) return `${hours} 小时前`;
  if (days < 7) return `${days} 天前`;
  return formatDate(ts);
}

export function groupTasksByDate(tasks: Task[]): TaskGroup[] {
  const now = Date.now();
  const today = startOfDay(now);
  const tomorrow = today + dayMs;
  const dayOfWeek = new Date(today).getDay();
  const diffToMonday = dayOfWeek === 0 ? 6 : dayOfWeek - 1;
  const weekStart = today - diffToMonday * dayMs;
  const weekEnd = weekStart + 7 * dayMs;

  const todayTasks: Task[] = [];
  const tomorrowTasks: Task[] = [];
  const weekTasks: Task[] = [];
  const laterTasks: Task[] = [];

  for (const task of tasks) {
    if (task.parentId) continue;
    if (!task.dueDate) {
      laterTasks.push(task);
      continue;
    }
    const due = startOfDay(task.dueDate);
    if (due === today) {
      todayTasks.push(task);
    } else if (due === tomorrow) {
      tomorrowTasks.push(task);
    } else if (due >= weekStart && due < weekEnd) {
      weekTasks.push(task);
    } else {
      laterTasks.push(task);
    }
  }

  const groups: TaskGroup[] = [];
  if (todayTasks.length > 0) {
    groups.push({ key: 'today', label: '今天', tasks: todayTasks.sort((a, b) => a.order - b.order) });
  }
  if (tomorrowTasks.length > 0) {
    groups.push({ key: 'tomorrow', label: '明天', tasks: tomorrowTasks.sort((a, b) => a.order - b.order) });
  }
  if (weekTasks.length > 0) {
    groups.push({ key: 'week', label: '本周', tasks: weekTasks.sort((a, b) => (a.dueDate || 0) - (b.dueDate || 0)) });
  }
  if (laterTasks.length > 0) {
    groups.push({ key: 'later', label: '更晚', tasks: laterTasks.sort((a, b) => (a.dueDate || Infinity) - (b.dueDate || Infinity)) });
  }

  return groups;
}

export function filterTasks(tasks: Task[], filter: FilterType, searchQuery: string): Task[] {
  let result = tasks;

  if (searchQuery.trim()) {
    const q = searchQuery.toLowerCase();
    result = result.filter(
      (t) =>
        t.title.toLowerCase().includes(q) ||
        (t.description && t.description.toLowerCase().includes(q)) ||
        (t.children && t.children.some((c) => c.title.toLowerCase().includes(q)))
    );
  }

  if (filter === 'all') return result;

  const now = Date.now();
  const today = startOfDay(now);
  const tomorrow = today + dayMs;
  const dayOfWeek = new Date(today).getDay();
  const diffToMonday = dayOfWeek === 0 ? 6 : dayOfWeek - 1;
  const weekStart = today - diffToMonday * dayMs;
  const weekEnd = weekStart + 7 * dayMs;

  return result.filter((task) => {
    if (!task.dueDate) return false;
    const due = startOfDay(task.dueDate);
    if (filter === 'today') return due === today;
    if (filter === 'tomorrow') return due === tomorrow;
    if (filter === 'week') return due >= weekStart && due < weekEnd;
    return true;
  });
}

export function calcProgress(task: Task): { completed: number; total: number; percent: number } {
  if (!task.children || task.children.length === 0) {
    const completed = task.status === 'completed' ? 1 : 0;
    return { completed, total: 1, percent: completed * 100 };
  }
  const total = task.children.length;
  const completed = task.children.filter((c) => c.status === 'completed').length;
  return { completed, total, percent: total > 0 ? Math.round((completed / total) * 100) : 0 };
}
